EpikCord/client/http.py

```diff
@@ -0,0 +1,475 @@
+            The response of the request.
+
+        Raises
+        ------
+        NotFound
+            If the request returns a 404 status code.
+        Forbidden
+            If the request returns a 403 status code.
+        Unauthorized
+            If the request returns a 401 status code.
+        BadRequest
+            If the request returns a 400 status code.
+        HTTPException
+            If the request returns a status code that is not OK and not any of the other exceptions.
+        """
+
+        method = route.method
+        url = route.url
+
+        if not discord:
+            return await self.session.request(method, url, *args, **kwargs)
+
+        self.setup_kwargs(kwargs, files=files, json=json)
+
+        url = self.clean_url(url)
+        bucket = self.buckets.get(f"{method}:{url}") or MockBucket()
+
+        await self.global_event.wait()
+        await bucket.wait()
+
+        for _ in range(5):
+
+            async with self.session.request(method, url, *args, **kwargs) as response:
+                if isinstance(bucket, MockBucket):
+                    bucket = await self.set_bucket(
+                        response=response,
+                        **route.major_parameters,
+                    )
+
+                data = await self.extract_content(response)
+
+                if response.status == 429:
+                    await self.handle_ratelimit(data, bucket)
+                elif response.headers.get("X-RateLimit-Remaining", "1") == "0":
+                    await self.handle_exhausted_bucket(response, bucket)
+                elif not response.ok:
+                    error = self.error_mapping.get(response.status, HTTPException)
+                    raise error(response, data)
+                elif response.ok:
+                    return response
+        raise TimeoutError("Attempted request 5 times .")
+
+    async def set_bucket(
+        self,
+        *,
+        response: aiohttp.ClientResponse,
+        channel_id: Optional[int] = None,
+        guild_id: Optional[int] = None,
+        webhook_id: Optional[int] = None,
+        webhook_token: Optional[str] = None,
+    ) -> Union[Bucket, TopLevelBucket]:
+        url = response.url
+        method = response.method
+
+        if channel_id or guild_id or webhook_id or webhook_token:
+            bucket = TopLevelBucket(
+                channel_id=channel_id,
+                guild_id=guild_id,
+                webhook_id=webhook_id,
+                webhook_token=webhook_token,
+                )
+
+        else:
+
+            bucket = Bucket(bucket_hash=response.headers["X-RateLimit-Bucket"])
+            if bucket in self.buckets.values():
+                self.buckets[f"{method}:{url}"] = list(self.buckets.values())[
+                    list(self.buckets.values()).index(bucket)
+                ]
+                bucket = self.buckets[f"{method}:{url}"]
+            self.buckets[f"{method}:{url}"] = bucket
```

Comment:
To remove some redundant code, you should make a variable named `bucket_key` that's set to `f"{method}:{url}"`.

Reply:
Yeah that's a good point you've made